lib/keycard/api_key.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require "digest"
+require "securerandom"
+
+# A typical api key, ready to be encrypted.
+class Keycard::ApiKey
+  class HiddenKeyError < StandardError; end
+
+  # @param key [String] An optional value for the key
+  def initialize(key = SecureRandom.uuid)
+    @key = key
+  end
+
+  # The key's value
+  def to_s
+    key
+  end
+
+  # The result of hasing the key
+  def digest
+    Digest::SHA256.hexdigest(key)
+  end
```

Comment:
I should say, the Rails custom type / attribute is sort of a prospective use case, but one I think could be pretty helpful. KCV exists to be convenience glue between Rails and Keycard (etc), so binding a generic facility like the keys to an AR type there seems appropriate. If you have an app where a model should hold an API key, it would be easy to add an attribute there with minimal code.

In my mind, this is a much better approach than the generic model mixins and DSL methods that do all kinds of metaprogramming in your model. Registering an attribute type that can be used with a standard (documented) extension mechanism is discoverable in a way that "acts_as_..." or "has_attachment" is not.

Reply:
I've added these changes. Take a look at the rails attribute stuff, specifically where I've chosen to conditionally subclass from ActiveRecord::Type::Value. The way they've chosen to do things is bizarre.

Reply:
Agreed. Per discussion today, the Type / Value / attribute stuff is tabled for now.